{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/tamucc-gcl/wrkshp_edna_metabarcoding/blob/main/analysis/initialize_environment.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "B8JlmtFMBjs9",
      "metadata": {
        "id": "B8JlmtFMBjs9"
      },
      "source": [
        "# Post-Bioionformatic Data Processing\n",
        "\n",
        "This notebook will:\n",
        "\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "---\n",
        "## Initialize Environment"
      ],
      "metadata": {
        "id": "ltzoYG-KATXP"
      },
      "id": "ltzoYG-KATXP"
    },
    {
      "cell_type": "code",
      "source": [
        "#Set-up Computing Environment - Python\n",
        "\n",
        "## Mount Google Drive\n",
        "from google.colab import drive\n",
        "import os\n",
        "drive.mount('/content/drive')\n",
        "os.environ['COLAB'] = 'TRUE'\n",
        "\n",
        "## Install Linux programs needed\n",
        "!apt install libfribidi-dev libglpk-dev libharfbuzz-dev pandoc\n",
        "\n",
        "## Clone GitHub Repo and move into that repo\n",
        "local_path = '/content/edna_workshop'\n",
        "repo  = \"tamucc-gcl/wrkshp_edna_metabarcoding\"\n",
        "url = f\"https://github.com/{repo}.git\"\n",
        "!git clone {url} {local_path}\n",
        "\n",
        "## Setup R and move to local directory\n",
        "os.chdir(local_path)\n",
        "%reload_ext rpy2.ipython"
      ],
      "metadata": {
        "id": "G0FnmIJ3G-7P",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "adb4df26-d198-422d-899d-60a894e2b7da",
        "collapsed": true
      },
      "id": "G0FnmIJ3G-7P",
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/drive\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  libamd2 libbtf1 libcamd2 libccolamd2 libcholmod3\n",
            "  libcmark-gfm-extensions0.29.0.gfm.3 libcmark-gfm0.29.0.gfm.3 libcolamd2\n",
            "  libcxsparse3 libglpk40 libgmp-dev libgmpxx4ldbl libgraphblas-dev\n",
            "  libgraphblas6 libgraphite2-dev libharfbuzz-gobject0 libharfbuzz-icu0 libklu1\n",
            "  libldl2 libmetis5 libmongoose2 librbio2 libsliplu1 libspqr2\n",
            "  libsuitesparse-dev libsuitesparseconfig5 libumfpack5 pandoc-data\n",
            "Suggested packages:\n",
            "  libiodbc2-dev gmp-doc libgmp10-doc libmpfr-dev libgraphite2-utils\n",
            "  texlive-latex-recommended texlive-xetex texlive-luatex pandoc-citeproc\n",
            "  texlive-latex-extra context wkhtmltopdf librsvg2-bin groff ghc nodejs php\n",
            "  python ruby libjs-mathjax libjs-katex citation-style-language-styles\n",
            "The following NEW packages will be installed:\n",
            "  libamd2 libbtf1 libcamd2 libccolamd2 libcholmod3\n",
            "  libcmark-gfm-extensions0.29.0.gfm.3 libcmark-gfm0.29.0.gfm.3 libcolamd2\n",
            "  libcxsparse3 libfribidi-dev libglpk-dev libglpk40 libgmp-dev libgmpxx4ldbl\n",
            "  libgraphblas-dev libgraphblas6 libgraphite2-dev libharfbuzz-dev\n",
            "  libharfbuzz-gobject0 libharfbuzz-icu0 libklu1 libldl2 libmetis5 libmongoose2\n",
            "  librbio2 libsliplu1 libspqr2 libsuitesparse-dev libsuitesparseconfig5\n",
            "  libumfpack5 pandoc pandoc-data\n",
            "0 upgraded, 32 newly installed, 0 to remove and 34 not upgraded.\n",
            "Need to get 44.7 MB of archives.\n",
            "After this operation, 334 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu jammy/main amd64 libsuitesparseconfig5 amd64 1:5.10.1+dfsg-4build1 [10.4 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libamd2 amd64 1:5.10.1+dfsg-4build1 [21.6 kB]\n",
            "Get:3 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libbtf1 amd64 1:5.10.1+dfsg-4build1 [12.1 kB]\n",
            "Get:4 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcamd2 amd64 1:5.10.1+dfsg-4build1 [23.3 kB]\n",
            "Get:5 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libccolamd2 amd64 1:5.10.1+dfsg-4build1 [25.2 kB]\n",
            "Get:6 http://archive.ubuntu.com/ubuntu jammy/main amd64 libcolamd2 amd64 1:5.10.1+dfsg-4build1 [18.0 kB]\n",
            "Get:7 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libmetis5 amd64 5.1.0.dfsg-7build2 [181 kB]\n",
            "Get:8 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcholmod3 amd64 1:5.10.1+dfsg-4build1 [346 kB]\n",
            "Get:9 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcmark-gfm0.29.0.gfm.3 amd64 0.29.0.gfm.3-3 [115 kB]\n",
            "Get:10 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcmark-gfm-extensions0.29.0.gfm.3 amd64 0.29.0.gfm.3-3 [25.1 kB]\n",
            "Get:11 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libcxsparse3 amd64 1:5.10.1+dfsg-4build1 [70.8 kB]\n",
            "Get:12 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libfribidi-dev amd64 1.0.8-2ubuntu3.1 [64.8 kB]\n",
            "Get:13 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libglpk40 amd64 5.0-1 [361 kB]\n",
            "Get:14 http://archive.ubuntu.com/ubuntu jammy/main amd64 libgmpxx4ldbl amd64 2:6.2.1+dfsg-3ubuntu1 [9,580 B]\n",
            "Get:15 http://archive.ubuntu.com/ubuntu jammy/main amd64 libgmp-dev amd64 2:6.2.1+dfsg-3ubuntu1 [337 kB]\n",
            "Get:16 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libklu1 amd64 1:5.10.1+dfsg-4build1 [77.6 kB]\n",
            "Get:17 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libldl2 amd64 1:5.10.1+dfsg-4build1 [11.7 kB]\n",
            "Get:18 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libmongoose2 amd64 1:5.10.1+dfsg-4build1 [33.5 kB]\n",
            "Get:19 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libumfpack5 amd64 1:5.10.1+dfsg-4build1 [250 kB]\n",
            "Get:20 http://archive.ubuntu.com/ubuntu jammy/universe amd64 librbio2 amd64 1:5.10.1+dfsg-4build1 [26.6 kB]\n",
            "Get:21 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libspqr2 amd64 1:5.10.1+dfsg-4build1 [71.6 kB]\n",
            "Get:22 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libsliplu1 amd64 1:5.10.1+dfsg-4build1 [37.1 kB]\n",
            "Get:23 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libsuitesparse-dev amd64 1:5.10.1+dfsg-4build1 [1,058 kB]\n",
            "Get:24 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libglpk-dev amd64 5.0-1 [430 kB]\n",
            "Get:25 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libgraphblas6 amd64 6.1.4+dfsg-2 [20.1 MB]\n",
            "Get:26 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libgraphblas-dev amd64 6.1.4+dfsg-2 [54.4 kB]\n",
            "Get:27 http://archive.ubuntu.com/ubuntu jammy/main amd64 libgraphite2-dev amd64 1.3.14-1build2 [14.7 kB]\n",
            "Get:28 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libharfbuzz-icu0 amd64 2.7.4-1ubuntu3.2 [5,890 B]\n",
            "Get:29 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libharfbuzz-gobject0 amd64 2.7.4-1ubuntu3.2 [26.0 kB]\n",
            "Get:30 http://archive.ubuntu.com/ubuntu jammy-updates/main amd64 libharfbuzz-dev amd64 2.7.4-1ubuntu3.2 [554 kB]\n",
            "Get:31 http://archive.ubuntu.com/ubuntu jammy/universe amd64 pandoc-data all 2.9.2.1-3ubuntu2 [81.8 kB]\n",
            "Get:32 http://archive.ubuntu.com/ubuntu jammy/universe amd64 pandoc amd64 2.9.2.1-3ubuntu2 [20.3 MB]\n",
            "Fetched 44.7 MB in 5s (9,671 kB/s)\n",
            "Extracting templates from packages: 100%\n",
            "Selecting previously unselected package libsuitesparseconfig5:amd64.\n",
            "(Reading database ... 126101 files and directories currently installed.)\n",
            "Preparing to unpack .../00-libsuitesparseconfig5_1%3a5.10.1+dfsg-4build1_amd64.deb ...\n",
            "Unpacking libsuitesparseconfig5:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Selecting previously unselected package libamd2:amd64.\n",
            "Preparing to unpack .../01-libamd2_1%3a5.10.1+dfsg-4build1_amd64.deb ...\n",
            "Unpacking libamd2:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Selecting previously unselected package libbtf1:amd64.\n",
            "Preparing to unpack .../02-libbtf1_1%3a5.10.1+dfsg-4build1_amd64.deb ...\n",
            "Unpacking libbtf1:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Selecting previously unselected package libcamd2:amd64.\n",
            "Preparing to unpack .../03-libcamd2_1%3a5.10.1+dfsg-4build1_amd64.deb ...\n",
            "Unpacking libcamd2:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Selecting previously unselected package libccolamd2:amd64.\n",
            "Preparing to unpack .../04-libccolamd2_1%3a5.10.1+dfsg-4build1_amd64.deb ...\n",
            "Unpacking libccolamd2:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Selecting previously unselected package libcolamd2:amd64.\n",
            "Preparing to unpack .../05-libcolamd2_1%3a5.10.1+dfsg-4build1_amd64.deb ...\n",
            "Unpacking libcolamd2:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Selecting previously unselected package libmetis5:amd64.\n",
            "Preparing to unpack .../06-libmetis5_5.1.0.dfsg-7build2_amd64.deb ...\n",
            "Unpacking libmetis5:amd64 (5.1.0.dfsg-7build2) ...\n",
            "Selecting previously unselected package libcholmod3:amd64.\n",
            "Preparing to unpack .../07-libcholmod3_1%3a5.10.1+dfsg-4build1_amd64.deb ...\n",
            "Unpacking libcholmod3:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Selecting previously unselected package libcmark-gfm0.29.0.gfm.3:amd64.\n",
            "Preparing to unpack .../08-libcmark-gfm0.29.0.gfm.3_0.29.0.gfm.3-3_amd64.deb ...\n",
            "Unpacking libcmark-gfm0.29.0.gfm.3:amd64 (0.29.0.gfm.3-3) ...\n",
            "Selecting previously unselected package libcmark-gfm-extensions0.29.0.gfm.3:amd64.\n",
            "Preparing to unpack .../09-libcmark-gfm-extensions0.29.0.gfm.3_0.29.0.gfm.3-3_amd64.deb ...\n",
            "Unpacking libcmark-gfm-extensions0.29.0.gfm.3:amd64 (0.29.0.gfm.3-3) ...\n",
            "Selecting previously unselected package libcxsparse3:amd64.\n",
            "Preparing to unpack .../10-libcxsparse3_1%3a5.10.1+dfsg-4build1_amd64.deb ...\n",
            "Unpacking libcxsparse3:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Selecting previously unselected package libfribidi-dev:amd64.\n",
            "Preparing to unpack .../11-libfribidi-dev_1.0.8-2ubuntu3.1_amd64.deb ...\n",
            "Unpacking libfribidi-dev:amd64 (1.0.8-2ubuntu3.1) ...\n",
            "Selecting previously unselected package libglpk40:amd64.\n",
            "Preparing to unpack .../12-libglpk40_5.0-1_amd64.deb ...\n",
            "Unpacking libglpk40:amd64 (5.0-1) ...\n",
            "Selecting previously unselected package libgmpxx4ldbl:amd64.\n",
            "Preparing to unpack .../13-libgmpxx4ldbl_2%3a6.2.1+dfsg-3ubuntu1_amd64.deb ...\n",
            "Unpacking libgmpxx4ldbl:amd64 (2:6.2.1+dfsg-3ubuntu1) ...\n",
            "Selecting previously unselected package libgmp-dev:amd64.\n",
            "Preparing to unpack .../14-libgmp-dev_2%3a6.2.1+dfsg-3ubuntu1_amd64.deb ...\n",
            "Unpacking libgmp-dev:amd64 (2:6.2.1+dfsg-3ubuntu1) ...\n",
            "Selecting previously unselected package libklu1:amd64.\n",
            "Preparing to unpack .../15-libklu1_1%3a5.10.1+dfsg-4build1_amd64.deb ...\n",
            "Unpacking libklu1:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Selecting previously unselected package libldl2:amd64.\n",
            "Preparing to unpack .../16-libldl2_1%3a5.10.1+dfsg-4build1_amd64.deb ...\n",
            "Unpacking libldl2:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Selecting previously unselected package libmongoose2:amd64.\n",
            "Preparing to unpack .../17-libmongoose2_1%3a5.10.1+dfsg-4build1_amd64.deb ...\n",
            "Unpacking libmongoose2:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Selecting previously unselected package libumfpack5:amd64.\n",
            "Preparing to unpack .../18-libumfpack5_1%3a5.10.1+dfsg-4build1_amd64.deb ...\n",
            "Unpacking libumfpack5:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Selecting previously unselected package librbio2:amd64.\n",
            "Preparing to unpack .../19-librbio2_1%3a5.10.1+dfsg-4build1_amd64.deb ...\n",
            "Unpacking librbio2:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Selecting previously unselected package libspqr2:amd64.\n",
            "Preparing to unpack .../20-libspqr2_1%3a5.10.1+dfsg-4build1_amd64.deb ...\n",
            "Unpacking libspqr2:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Selecting previously unselected package libsliplu1:amd64.\n",
            "Preparing to unpack .../21-libsliplu1_1%3a5.10.1+dfsg-4build1_amd64.deb ...\n",
            "Unpacking libsliplu1:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Selecting previously unselected package libsuitesparse-dev:amd64.\n",
            "Preparing to unpack .../22-libsuitesparse-dev_1%3a5.10.1+dfsg-4build1_amd64.deb ...\n",
            "Unpacking libsuitesparse-dev:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Selecting previously unselected package libglpk-dev:amd64.\n",
            "Preparing to unpack .../23-libglpk-dev_5.0-1_amd64.deb ...\n",
            "Unpacking libglpk-dev:amd64 (5.0-1) ...\n",
            "Selecting previously unselected package libgraphblas6:amd64.\n",
            "Preparing to unpack .../24-libgraphblas6_6.1.4+dfsg-2_amd64.deb ...\n",
            "Unpacking libgraphblas6:amd64 (6.1.4+dfsg-2) ...\n",
            "Selecting previously unselected package libgraphblas-dev:amd64.\n",
            "Preparing to unpack .../25-libgraphblas-dev_6.1.4+dfsg-2_amd64.deb ...\n",
            "Unpacking libgraphblas-dev:amd64 (6.1.4+dfsg-2) ...\n",
            "Selecting previously unselected package libgraphite2-dev:amd64.\n",
            "Preparing to unpack .../26-libgraphite2-dev_1.3.14-1build2_amd64.deb ...\n",
            "Unpacking libgraphite2-dev:amd64 (1.3.14-1build2) ...\n",
            "Selecting previously unselected package libharfbuzz-icu0:amd64.\n",
            "Preparing to unpack .../27-libharfbuzz-icu0_2.7.4-1ubuntu3.2_amd64.deb ...\n",
            "Unpacking libharfbuzz-icu0:amd64 (2.7.4-1ubuntu3.2) ...\n",
            "Selecting previously unselected package libharfbuzz-gobject0:amd64.\n",
            "Preparing to unpack .../28-libharfbuzz-gobject0_2.7.4-1ubuntu3.2_amd64.deb ...\n",
            "Unpacking libharfbuzz-gobject0:amd64 (2.7.4-1ubuntu3.2) ...\n",
            "Selecting previously unselected package libharfbuzz-dev:amd64.\n",
            "Preparing to unpack .../29-libharfbuzz-dev_2.7.4-1ubuntu3.2_amd64.deb ...\n",
            "Unpacking libharfbuzz-dev:amd64 (2.7.4-1ubuntu3.2) ...\n",
            "Selecting previously unselected package pandoc-data.\n",
            "Preparing to unpack .../30-pandoc-data_2.9.2.1-3ubuntu2_all.deb ...\n",
            "Unpacking pandoc-data (2.9.2.1-3ubuntu2) ...\n",
            "Selecting previously unselected package pandoc.\n",
            "Preparing to unpack .../31-pandoc_2.9.2.1-3ubuntu2_amd64.deb ...\n",
            "Unpacking pandoc (2.9.2.1-3ubuntu2) ...\n",
            "Setting up libharfbuzz-icu0:amd64 (2.7.4-1ubuntu3.2) ...\n",
            "Setting up libfribidi-dev:amd64 (1.0.8-2ubuntu3.1) ...\n",
            "Setting up libgraphblas6:amd64 (6.1.4+dfsg-2) ...\n",
            "Setting up libldl2:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Setting up libmetis5:amd64 (5.1.0.dfsg-7build2) ...\n",
            "Setting up libharfbuzz-gobject0:amd64 (2.7.4-1ubuntu3.2) ...\n",
            "Setting up libbtf1:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Setting up libgraphblas-dev:amd64 (6.1.4+dfsg-2) ...\n",
            "Setting up libgraphite2-dev:amd64 (1.3.14-1build2) ...\n",
            "Setting up libgmpxx4ldbl:amd64 (2:6.2.1+dfsg-3ubuntu1) ...\n",
            "Setting up libcmark-gfm0.29.0.gfm.3:amd64 (0.29.0.gfm.3-3) ...\n",
            "Setting up libcmark-gfm-extensions0.29.0.gfm.3:amd64 (0.29.0.gfm.3-3) ...\n",
            "Setting up libcxsparse3:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Setting up pandoc-data (2.9.2.1-3ubuntu2) ...\n",
            "Setting up libsuitesparseconfig5:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Setting up librbio2:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Setting up libamd2:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Setting up libharfbuzz-dev:amd64 (2.7.4-1ubuntu3.2) ...\n",
            "Setting up libgmp-dev:amd64 (2:6.2.1+dfsg-3ubuntu1) ...\n",
            "Setting up libcolamd2:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Setting up libsliplu1:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Setting up libcamd2:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Setting up libmongoose2:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Setting up pandoc (2.9.2.1-3ubuntu2) ...\n",
            "Setting up libglpk40:amd64 (5.0-1) ...\n",
            "Setting up libklu1:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Setting up libccolamd2:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Setting up libcholmod3:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Setting up libspqr2:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Setting up libumfpack5:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Setting up libsuitesparse-dev:amd64 (1:5.10.1+dfsg-4build1) ...\n",
            "Setting up libglpk-dev:amd64 (5.0-1) ...\n",
            "Processing triggers for man-db (2.10.2-1) ...\n",
            "Processing triggers for libc-bin (2.35-0ubuntu3.8) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc_proxy.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libhwloc.so.15 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_level_zero.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_loader.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_5.so.3 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbmalloc.so.2 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libur_adapter_opencl.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtcm_debug.so.1 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libumf.so.0 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbb.so.12 is not a symbolic link\n",
            "\n",
            "/sbin/ldconfig.real: /usr/local/lib/libtbbbind_2_0.so.3 is not a symbolic link\n",
            "\n",
            "Cloning into '/content/edna_workshop'...\n",
            "remote: Enumerating objects: 290, done.\u001b[K\n",
            "remote: Counting objects: 100% (290/290), done.\u001b[K\n",
            "remote: Compressing objects: 100% (238/238), done.\u001b[K\n",
            "remote: Total 290 (delta 116), reused 193 (delta 41), pack-reused 0 (from 0)\u001b[K\n",
            "Receiving objects: 100% (290/290), 15.94 MiB | 19.83 MiB/s, done.\n",
            "Resolving deltas: 100% (116/116), done.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Set-up Computing Environment - R\n",
        "%%R -i local_path\n",
        "# Point R at your Drive-backed library\n",
        "drive_lib <- \"/content/drive/MyDrive/edna_libraries\"\n",
        "dir.create(drive_lib, recursive = TRUE, showWarnings = FALSE)\n",
        ".libPaths(c(drive_lib, .libPaths()))\n",
        "\n",
        "#Install renv if it isn't already\n",
        "if (!require(\"renv\", lib.loc = drive_lib, quietly = TRUE, warn.conflicts = FALSE)) {\n",
        "  install.packages(\"renv\", lib = drive_lib,\n",
        "                   quietly = TRUE, warn.conflicts = FALSE)\n",
        "}\n",
        "\n",
        "#Install required libraries into google drive if they aren't already\n",
        "#Sys.setenv(RENV_CONFIG_REPOS_OVERRIDE = \"https://cloud.r-project.org\")\n",
        "#Sys.setenv(RENV_CONFIG_REPOS_OVERRIDE = \"https://cloud.r-project.org@https://bioconductor.org/packages/3.19/bioc\")\n",
        "renv::restore(library = drive_lib,\n",
        "              lockfile = paste0(local_path, '/renv.lock'),\n",
        "              prompt = FALSE)\n",
        "\n",
        "#Create a persistent storage temporary file directory\n",
        "persistent_directory <- \"/content/drive/MyDrive/edna_libraries/intermediate_files\"\n",
        "dir.create(persistent_directory, recursive = TRUE, showWarnings = FALSE)"
      ],
      "metadata": {
        "id": "N3uZ6R1lHhaB",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "e4a61b8e-dab8-4cb6-c161-a8061b02e16d",
        "collapsed": true
      },
      "id": "N3uZ6R1lHhaB",
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "The following package(s) will be updated:\n",
            "\n",
            "# Bioconductor ---------------------------------------------------------------\n",
            "- GenomeInfoDbData        [* -> 1.2.13]\n",
            "\n",
            "# Bioconductor 3.20 ----------------------------------------------------------\n",
            "- BiocGenerics            [* -> 0.52.0]\n",
            "- BiocVersion             [* -> 3.20.0]\n",
            "- Biostrings              [* -> 2.74.1]\n",
            "- GenomeInfoDb            [* -> 1.42.3]\n",
            "- IRanges                 [* -> 2.40.1]\n",
            "- S4Vectors               [* -> 0.44.0]\n",
            "- UCSC.utils              [* -> 1.2.0]\n",
            "- XVector                 [* -> 0.46.0]\n",
            "- zlibbioc                [* -> 1.52.0]\n",
            "\n",
            "# CRAN -----------------------------------------------------------------------\n",
            "- ape                     [* -> 5.8-1]\n",
            "- askpass                 [* -> 1.2.1]\n",
            "- assertthat              [* -> 0.2.1]\n",
            "- backports               [* -> 1.5.0]\n",
            "- base64enc               [* -> 0.1-3]\n",
            "- bit                     [* -> 4.6.0]\n",
            "- bit64                   [* -> 4.6.0-1]\n",
            "- blob                    [* -> 1.2.4]\n",
            "- broom                   [* -> 1.0.8]\n",
            "- bslib                   [* -> 0.9.0]\n",
            "- cachem                  [* -> 1.1.0]\n",
            "- callr                   [* -> 3.7.6]\n",
            "- cellranger              [* -> 1.1.0]\n",
            "- checkmate               [* -> 2.3.2]\n",
            "- cli                     [* -> 3.6.5]\n",
            "- clipr                   [* -> 0.8.0]\n",
            "- cluster                 [* -> 2.1.6]\n",
            "- colorspace              [* -> 2.1-1]\n",
            "- commonmark              [* -> 1.9.5]\n",
            "- conflicted              [* -> 1.2.0]\n",
            "- cpp11                   [* -> 0.5.2]\n",
            "- crayon                  [* -> 1.5.3]\n",
            "- crul                    [* -> 1.5.0]\n",
            "- curl                    [* -> 6.2.2]\n",
            "- data.table              [* -> 1.17.0]\n",
            "- DBI                     [* -> 1.2.3]\n",
            "- dbplyr                  [* -> 2.5.0]\n",
            "- digest                  [* -> 0.6.37]\n",
            "- dplyr                   [* -> 1.1.4]\n",
            "- dtplyr                  [* -> 1.3.1]\n",
            "- duckdb                  [* -> 1.2.1]\n",
            "- duckdbfs                [* -> 0.1.0]\n",
            "- evaluate                [* -> 1.0.3]\n",
            "- fansi                   [* -> 1.0.6]\n",
            "- farver                  [* -> 2.1.2]\n",
            "- fastmap                 [* -> 1.2.0]\n",
            "- fastmatch               [* -> 1.1-6]\n",
            "- fontawesome             [* -> 0.5.3]\n",
            "- forcats                 [* -> 1.0.0]\n",
            "- foreign                 [* -> 0.8-86]\n",
            "- Formula                 [* -> 1.2-5]\n",
            "- fs                      [* -> 1.6.6]\n",
            "- gargle                  [* -> 1.5.2]\n",
            "- generics                [* -> 0.1.3]\n",
            "- ggplot2                 [* -> 3.5.2]\n",
            "- ggrepel                 [* -> 0.9.6]\n",
            "- ggtext                  [* -> 0.1.2]\n",
            "- glue                    [* -> 1.8.0]\n",
            "- googledrive             [* -> 2.1.1]\n",
            "- googlesheets4           [* -> 1.1.1]\n",
            "- gridExtra               [* -> 2.3]\n",
            "- gridtext                [* -> 0.1.5]\n",
            "- gtable                  [* -> 0.3.6]\n",
            "- haven                   [* -> 2.5.4]\n",
            "- here                    [* -> 1.0.1]\n",
            "- highr                   [* -> 0.11]\n",
            "- Hmisc                   [* -> 5.2-3]\n",
            "- hms                     [* -> 1.1.3]\n",
            "- htmlTable               [* -> 2.4.3]\n",
            "- htmltools               [* -> 0.5.8.1]\n",
            "- htmlwidgets             [* -> 1.6.4]\n",
            "- httpcode                [* -> 0.3.0]\n",
            "- httr                    [* -> 1.4.7]\n",
            "- ids                     [* -> 1.0.1]\n",
            "- igraph                  [* -> 2.1.4]\n",
            "- isoband                 [* -> 0.2.7]\n",
            "- janitor                 [* -> 2.2.1]\n",
            "- jpeg                    [* -> 0.1-11]\n",
            "- jquerylib               [* -> 0.1.4]\n",
            "- jsonlite                [* -> 2.0.0]\n",
            "- knitr                   [* -> 1.50]\n",
            "- labeling                [* -> 0.4.3]\n",
            "- later                   [* -> 1.4.2]\n",
            "- lattice                 [* -> 0.22-6]\n",
            "- lazyeval                [* -> 0.2.2]\n",
            "- lifecycle               [* -> 1.0.4]\n",
            "- litedown                [* -> 0.7]\n",
            "- lubridate               [* -> 1.9.4]\n",
            "- magrittr                [* -> 2.0.3]\n",
            "- markdown                [* -> 2.0]\n",
            "- MASS                    [* -> 7.3-60.2]\n",
            "- Matrix                  [* -> 1.7-0]\n",
            "- memoise                 [* -> 2.0.1]\n",
            "- mgcv                    [* -> 1.9-1]\n",
            "- mime                    [* -> 0.13]\n",
            "- modelr                  [* -> 0.1.11]\n",
            "- natserv                 [* -> 1.0.0]\n",
            "- nlme                    [* -> 3.1-164]\n",
            "- nnet                    [* -> 7.3-19]\n",
            "- openssl                 [* -> 2.3.2]\n",
            "- patchwork               [* -> 1.3.0]\n",
            "- pbapply                 [* -> 1.7-2]\n",
            "- pillar                  [* -> 1.10.2]\n",
            "- pkgconfig               [* -> 2.0.3]\n",
            "- plyr                    [* -> 1.8.9]\n",
            "- png                     [* -> 0.1-8]\n",
            "- prettyunits             [* -> 1.2.0]\n",
            "- processx                [* -> 3.8.6]\n",
            "- progress                [* -> 1.2.3]\n",
            "- ps                      [* -> 1.9.1]\n",
            "- purrr                   [* -> 1.0.4]\n",
            "- quadprog                [* -> 1.5-8]\n",
            "- R6                      [* -> 2.6.1]\n",
            "- ragg                    [* -> 1.4.0]\n",
            "- rappdirs                [* -> 0.3.3]\n",
            "- ratelimitr              [* -> 0.4.1]\n",
            "- RColorBrewer            [* -> 1.1-3]\n",
            "- Rcpp                    [* -> 1.0.14]\n",
            "- readr                   [* -> 2.1.5]\n",
            "- readxl                  [* -> 1.4.5]\n",
            "- rematch                 [* -> 2.0.0]\n",
            "- rematch2                [* -> 2.1.2]\n",
            "- rentrez                 [* -> 1.2.3]\n",
            "- reprex                  [* -> 2.1.1]\n",
            "- rex                     [* -> 1.2.1]\n",
            "- rfishbase               [* -> 5.0.1]\n",
            "- ritis                   [* -> 1.0.0]\n",
            "- rlang                   [* -> 1.1.6]\n",
            "- rmarkdown               [* -> 2.29]\n",
            "- rncl                    [* -> 0.8.7]\n",
            "- rotl                    [* -> 3.1.0]\n",
            "- rpart                   [* -> 4.1.23]\n",
            "- rprojroot               [* -> 2.0.4]\n",
            "- rredlist                [* -> 1.0.0]\n",
            "- rstudioapi              [* -> 0.17.1]\n",
            "- rvest                   [* -> 1.0.4]\n",
            "- sass                    [* -> 0.4.10]\n",
            "- scales                  [* -> 1.4.0]\n",
            "- selectr                 [* -> 0.4-2]\n",
            "- snakecase               [* -> 0.11.1]\n",
            "- solrium                 [* -> 1.2.0]\n",
            "- stringi                 [* -> 1.8.7]\n",
            "- stringr                 [* -> 1.5.1]\n",
            "- sys                     [* -> 3.4.3]\n",
            "- systemfonts             [* -> 1.2.2]\n",
            "- taxize                  [* -> 0.10.0]\n",
            "- textshaping             [* -> 1.0.0]\n",
            "- tibble                  [* -> 3.2.1]\n",
            "- tidyr                   [* -> 1.3.1]\n",
            "- tidyselect              [* -> 1.2.1]\n",
            "- tidyverse               [* -> 2.0.0]\n",
            "- timechange              [* -> 0.3.0]\n",
            "- tinytex                 [* -> 0.57]\n",
            "- triebeard               [* -> 0.4.1]\n",
            "- tzdb                    [* -> 0.5.0]\n",
            "- urltools                [* -> 1.7.3]\n",
            "- utf8                    [* -> 1.2.4]\n",
            "- uuid                    [* -> 1.2-1]\n",
            "- vctrs                   [* -> 0.6.5]\n",
            "- viridis                 [* -> 0.6.5]\n",
            "- viridisLite             [* -> 0.4.2]\n",
            "- vroom                   [* -> 1.6.5]\n",
            "- WikidataQueryServiceR   [* -> 1.0.0]\n",
            "- WikidataR               [* -> 2.3.3]\n",
            "- WikipediR               [* -> 1.7.1]\n",
            "- wikitaxa                [* -> 0.4.0]\n",
            "- withr                   [* -> 3.0.2]\n",
            "- worrms                  [* -> 0.4.3]\n",
            "- xfun                    [* -> 0.52]\n",
            "- XML                     [* -> 3.99-0.18]\n",
            "- xml2                    [* -> 1.3.8]\n",
            "- zoo                     [* -> 1.8-14]\n",
            "\n",
            "# RSPM -----------------------------------------------------------------------\n",
            "- phangorn                [* -> 2.12.1]\n",
            "\n",
            "# https://packagemanager.posit.co/cran/2024-07-29 ----------------------------\n",
            "- bold                    [* -> 1.3.0]\n",
            "\n",
            "# Downloading packages -------------------------------------------------------\n",
            "- Downloading BiocGenerics from Bioconductor 3.20 ... OK [55.7 Kb in 0.63s]\n",
            "- Downloading BiocVersion from Bioconductor 3.20 ... OK [983 bytes in 0.54s]\n",
            "- Downloading Biostrings from Bioconductor 3.20 ... OK [12.2 Mb in 0.38s]\n",
            "- Downloading S4Vectors from Bioconductor 3.20 ... OK [1 Mb in 0.45s]\n",
            "- Downloading IRanges from Bioconductor 3.20 ... OK [475.6 Kb in 0.31s]\n",
            "- Downloading XVector from Bioconductor 3.20 ... OK [68.5 Kb in 0.41s]\n",
            "- Downloading zlibbioc from Bioconductor 3.20 ... OK [375.8 Kb in 0.48s]\n",
            "- Downloading GenomeInfoDb from Bioconductor 3.20 ... OK [3.4 Mb in 0.48s]\n",
            "- Downloading UCSC.utils from Bioconductor 3.20 ... OK [231.3 Kb in 0.66s]\n",
            "- Downloading GenomeInfoDbData from Bioconductor ... OK [12.3 Mb in 0.63s]\n",
            "- Downloading DBI from CRAN ...                 OK [898 Kb in 2.6s]\n",
            "- Downloading Formula from CRAN ...             OK [155.5 Kb in 1.9s]\n",
            "- Downloading Hmisc from CRAN ...               OK [3.5 Mb in 2.9s]\n",
            "- Downloading ggplot2 from CRAN ...             OK [4.8 Mb in 3.0s]\n",
            "- Downloading cli from CRAN ...                 OK [1.3 Mb in 2.5s]\n",
            "- Downloading glue from CRAN ...                OK [164.4 Kb in 1.9s]\n",
            "- Downloading gtable from CRAN ...              OK [217.4 Kb in 2.1s]\n",
            "- Downloading lifecycle from CRAN ...           OK [122.2 Kb in 1.9s]\n",
            "- Downloading rlang from CRAN ...               OK [1.5 Mb in 2.5s]\n",
            "- Downloading isoband from CRAN ...             OK [1.6 Mb in 2.7s]\n",
            "- Downloading MASS from CRAN ...                OK [498.9 Kb in 2.4s]\n",
            "- Downloading mgcv from CRAN ...                OK [1 Mb in 2.6s]\n",
            "- Downloading Matrix from CRAN ...              OK [2.4 Mb in 2.9s]\n",
            "- Downloading lattice from CRAN ...             OK [586.5 Kb in 2.3s]\n",
            "- Downloading nlme from CRAN ...                OK [818.7 Kb in 2.6s]\n",
            "- Downloading scales from CRAN ...              OK [822 Kb in 2.5s]\n",
            "- Downloading farver from CRAN ...              OK [1.4 Mb in 2.6s]\n",
            "- Downloading labeling from CRAN ...            OK [59.5 Kb in 1.7s]\n",
            "- Downloading RColorBrewer from CRAN ...        OK [50.8 Kb in 1.7s]\n",
            "- Downloading viridisLite from CRAN ...         OK [1.2 Mb in 2.5s]\n",
            "- Downloading tibble from CRAN ...              OK [663 Kb in 2.3s]\n",
            "- Downloading fansi from CRAN ...               OK [309.9 Kb in 2.1s]\n",
            "- Downloading magrittr from CRAN ...            OK [218.6 Kb in 2.2s]\n",
            "- Downloading pillar from CRAN ...              OK [644.3 Kb in 2.3s]\n",
            "- Downloading utf8 from CRAN ...                OK [144.6 Kb in 1.9s]\n",
            "- Downloading vctrs from CRAN ...               OK [1.3 Mb in 2.5s]\n",
            "- Downloading pkgconfig from CRAN ...           OK [17.7 Kb in 1.5s]\n",
            "- Downloading withr from CRAN ...               OK [218.6 Kb in 2.2s]\n",
            "- Downloading cluster from CRAN ...             OK [359.7 Kb in 2.2s]\n",
            "- Downloading rpart from CRAN ...               OK [605.7 Kb in 2.4s]\n",
            "- Downloading nnet from CRAN ...                OK [28.7 Kb in 1.7s]\n",
            "- Downloading foreign from CRAN ...             OK [355.7 Kb in 2.1s]\n",
            "- Downloading gridExtra from CRAN ...           OK [1.1 Mb in 2.5s]\n",
            "- Downloading data.table from CRAN ...          OK [2.6 Mb in 2.7s]\n",
            "- Downloading htmlTable from CRAN ...           OK [407.7 Kb in 2.3s]\n",
            "- Downloading stringr from CRAN ...             OK [303.7 Kb in 2.1s]\n",
            "- Downloading stringi from CRAN ...             OK [3.2 Mb in 3.0s]\n",
            "- Downloading knitr from CRAN ...               OK [1 Mb in 2.6s]\n",
            "- Downloading evaluate from CRAN ...            OK [99.2 Kb in 1.9s]\n",
            "- Downloading highr from CRAN ...               OK [36.8 Kb in 1.7s]\n",
            "- Downloading xfun from CRAN ...                OK [579.3 Kb in 2.3s]\n",
            "- Downloading checkmate from CRAN ...           OK [726.3 Kb in 2.3s]\n",
            "- Downloading backports from CRAN ...           OK [112.6 Kb in 1.9s]\n",
            "- Downloading htmlwidgets from CRAN ...         OK [796.3 Kb in 2.5s]\n",
            "- Downloading htmltools from CRAN ...           OK [347 Kb in 2.1s]\n",
            "- Downloading base64enc from CRAN ...           OK [26 Kb in 1.7s]\n",
            "- Downloading digest from CRAN ...              OK [217.7 Kb in 2.1s]\n",
            "- Downloading fastmap from CRAN ...             OK [61.8 Kb in 1.7s]\n",
            "- Downloading rmarkdown from CRAN ...           OK [2.5 Mb in 2.7s]\n",
            "- Downloading bslib from CRAN ...               OK [5.4 Mb in 2.9s]\n",
            "- Downloading cachem from CRAN ...              OK [66 Kb in 1.7s]\n",
            "- Downloading jquerylib from CRAN ...           OK [514.5 Kb in 2.3s]\n",
            "- Downloading memoise from CRAN ...             OK [47.7 Kb in 1.7s]\n",
            "- Downloading sass from CRAN ...                OK [2.3 Mb in 2.7s]\n",
            "- Downloading fs from CRAN ...                  OK [302.8 Kb in 2.2s]\n",
            "- Downloading rappdirs from CRAN ...            OK [44.3 Kb in 1.7s]\n",
            "- Downloading fontawesome from CRAN ...         OK [1.3 Mb in 2.5s]\n",
            "- Downloading tinytex from CRAN ...             OK [140.4 Kb in 1.9s]\n",
            "- Downloading rstudioapi from CRAN ...          OK [310.4 Kb in 2.1s]\n",
            "- Downloading viridis from CRAN ...             OK [2.9 Mb in 2.8s]\n",
            "- Downloading colorspace from CRAN ...          OK [2.5 Mb in 2.7s]\n",
            "- Downloading Rcpp from CRAN ...                OK [2.1 Mb in 2.7s]\n",
            "- Downloading WikidataQueryServiceR from CRAN ... OK [24.3 Kb in 1.7s]\n",
            "- Downloading dplyr from CRAN ...               OK [1.4 Mb in 2.5s]\n",
            "- Downloading generics from CRAN ...            OK [80.8 Kb in 2.0s]\n",
            "- Downloading tidyselect from CRAN ...          OK [220.2 Kb in 2.1s]\n",
            "- Downloading WikipediR from CRAN ...           OK [97.7 Kb in 1.9s]\n",
            "- Downloading ratelimitr from CRAN ...          OK [40.9 Kb in 1.7s]\n",
            "- Downloading assertthat from CRAN ...          OK [52.3 Kb in 1.7s]\n",
            "- Downloading purrr from CRAN ...               OK [524.8 Kb in 2.3s]\n",
            "- Downloading readr from CRAN ...               OK [841.4 Kb in 2.6s]\n",
            "- Downloading clipr from CRAN ...               OK [50.1 Kb in 1.7s]\n",
            "- Downloading hms from CRAN ...                 OK [97.8 Kb in 1.9s]\n",
            "- Downloading vroom from CRAN ...               OK [925.1 Kb in 2.5s]\n",
            "- Downloading bit64 from CRAN ...               OK [481 Kb in 2.3s]\n",
            "- Downloading bit from CRAN ...                 OK [616 Kb in 2.3s]\n",
            "- Downloading tzdb from CRAN ...                OK [648 Kb in 2.4s]\n",
            "- Downloading cpp11 from CRAN ...               OK [282.8 Kb in 2.1s]\n",
            "- Downloading progress from CRAN ...            OK [85 Kb in 1.9s]\n",
            "- Downloading prettyunits from CRAN ...         OK [149.2 Kb in 1.9s]\n",
            "- Downloading rex from CRAN ...                 OK [123.8 Kb in 1.9s]\n",
            "- Downloading lazyeval from CRAN ...            OK [153.5 Kb in 2.0s]\n",
            "- Downloading WikidataR from CRAN ...           OK [1.1 Mb in 2.5s]\n",
            "- Downloading pbapply from CRAN ...             OK [98.1 Kb in 1.9s]\n",
            "- Downloading XML from CRAN ...                 OK [1.7 Mb in 2.6s]\n",
            "- Downloading ape from CRAN ...                 OK [2.8 Mb in 2.7s]\n",
            "- Downloading blob from CRAN ...                OK [46.6 Kb in 1.7s]\n",
            "- Querying repositories for available source packages ... Done!\n",
            "- Downloading bold from https://packagemanager.posit.co/cran/2024-07-29 ... OK [1.9 Mb in 2.7s]\n",
            "- Downloading xml2 from CRAN ...                OK [275.2 Kb in 2.1s]\n",
            "- Downloading crul from CRAN ...                OK [652.7 Kb in 2.5s]\n",
            "- Downloading urltools from CRAN ...            OK [300 Kb in 2.2s]\n",
            "- Downloading triebeard from CRAN ...           OK [190.1 Kb in 2.1s]\n",
            "- Downloading httpcode from CRAN ...            OK [31.9 Kb in 1.7s]\n",
            "- Downloading broom from CRAN ...               OK [1.8 Mb in 2.7s]\n",
            "- Downloading tidyr from CRAN ...               OK [1.1 Mb in 2.4s]\n",
            "- Downloading callr from CRAN ...               OK [438.9 Kb in 2.4s]\n",
            "- Downloading processx from CRAN ...            OK [329.4 Kb in 2.1s]\n",
            "- Downloading ps from CRAN ...                  OK [488.8 Kb in 2.3s]\n",
            "- Downloading cellranger from CRAN ...          OK [101.5 Kb in 1.9s]\n",
            "- Downloading rematch from CRAN ...             OK [15.9 Kb in 1.5s]\n",
            "- Downloading commonmark from CRAN ...          OK [144 Kb in 2.0s]\n",
            "- Downloading conflicted from CRAN ...          OK [53.8 Kb in 1.8s]\n",
            "- Downloading dbplyr from CRAN ...              OK [1.2 Mb in 2.5s]\n",
            "- Downloading dtplyr from CRAN ...              OK [349.5 Kb in 2.1s]\n",
            "- Downloading duckdb from CRAN ...              OK [13.3 Mb in 2.3s]\n",
            "- Downloading duckdbfs from CRAN ...            OK [553.5 Kb in 2.3s]\n",
            "- Downloading fastmatch from CRAN ...           OK [35.2 Kb in 1.8s]\n",
            "- Downloading forcats from CRAN ...             OK [411.7 Kb in 2.3s]\n",
            "- Downloading gargle from CRAN ...              OK [757.5 Kb in 2.3s]\n",
            "- Downloading ggrepel from CRAN ...             OK [282.6 Kb in 2.1s]\n",
            "- Downloading ggtext from CRAN ...              OK [1.2 Mb in 2.5s]\n",
            "- Downloading gridtext from CRAN ...            OK [346.6 Kb in 2.1s]\n",
            "- Downloading markdown from CRAN ...            OK [62.8 Kb in 1.8s]\n",
            "- Downloading litedown from CRAN ...            OK [362.6 Kb in 2.3s]\n",
            "- Downloading png from CRAN ...                 OK [39.6 Kb in 1.7s]\n",
            "- Downloading jpeg from CRAN ...                OK [29.3 Kb in 1.7s]\n",
            "- Downloading googledrive from CRAN ...         OK [1.8 Mb in 2.7s]\n",
            "- Downloading uuid from CRAN ...                OK [47.7 Kb in 1.7s]\n",
            "- Downloading googlesheets4 from CRAN ...       OK [504.9 Kb in 2.4s]\n",
            "- Downloading ids from CRAN ...                 OK [119.8 Kb in 2.0s]\n",
            "- Downloading rematch2 from CRAN ...            OK [45 Kb in 1.7s]\n",
            "- Downloading haven from CRAN ...               OK [375.2 Kb in 2.1s]\n",
            "- Downloading here from CRAN ...                OK [52.2 Kb in 1.7s]\n",
            "- Downloading rprojroot from CRAN ...           OK [104.5 Kb in 2.0s]\n",
            "- Downloading igraph from CRAN ...              OK [5.3 Mb in 3.1s]\n",
            "- Downloading janitor from CRAN ...             OK [281.4 Kb in 2.1s]\n",
            "- Downloading lubridate from CRAN ...           OK [969.9 Kb in 2.5s]\n",
            "- Downloading timechange from CRAN ...          OK [166.5 Kb in 1.9s]\n",
            "- Downloading snakecase from CRAN ...           OK [157.9 Kb in 1.9s]\n",
            "- Downloading later from CRAN ...               OK [152 Kb in 2.0s]\n",
            "- Downloading modelr from CRAN ...              OK [196.1 Kb in 2.2s]\n",
            "- Downloading natserv from CRAN ...             OK [145.8 Kb in 1.9s]\n",
            "- Downloading patchwork from CRAN ...           OK [3.2 Mb in 2.9s]\n",
            "- Downloading phangorn from RSPM ...            OK [2.6 Mb in 2.9s]\n",
            "- Downloading quadprog from CRAN ...            OK [30.1 Kb in 1.7s]\n",
            "- Downloading plyr from CRAN ...                OK [768.3 Kb in 2.4s]\n",
            "- Downloading ragg from CRAN ...                OK [646.1 Kb in 2.3s]\n",
            "- Downloading systemfonts from CRAN ...         OK [321.3 Kb in 2.2s]\n",
            "- Downloading textshaping from CRAN ...         OK [179.1 Kb in 2.1s]\n",
            "- Downloading readxl from CRAN ...              OK [398.9 Kb in 2.3s]\n",
            "- Downloading rentrez from CRAN ...             OK [144.3 Kb in 2.1s]\n",
            "- Downloading reprex from CRAN ...              OK [483.4 Kb in 2.4s]\n",
            "- Downloading rfishbase from CRAN ...           OK [802.2 Kb in 2.5s]\n",
            "- Downloading ritis from CRAN ...               OK [151.3 Kb in 1.9s]\n",
            "- Downloading solrium from CRAN ...             OK [417.7 Kb in 2.3s]\n",
            "- Downloading rncl from CRAN ...                OK [597.5 Kb in 2.3s]\n",
            "- Downloading rotl from CRAN ...                OK [561.5 Kb in 2.4s]\n",
            "- Downloading rredlist from CRAN ...            OK [508.8 Kb in 2.3s]\n",
            "- Downloading rvest from CRAN ...               OK [293.1 Kb in 2.3s]\n",
            "- Downloading selectr from CRAN ...             OK [491.3 Kb in 2.4s]\n",
            "- Downloading taxize from CRAN ...              OK [1.5 Mb in 2.7s]\n",
            "- Downloading zoo from CRAN ...                 OK [1008 Kb in 2.6s]\n",
            "- Downloading worrms from CRAN ...              OK [101.5 Kb in 1.9s]\n",
            "- Downloading wikitaxa from CRAN ...            OK [438.1 Kb in 2.3s]\n",
            "- Downloading tidyverse from CRAN ...           OK [415.9 Kb in 2.3s]\n",
            "Successfully downloaded 165 packages in 400 seconds.\n",
            "\n",
            "# Installing packages --------------------------------------------------------\n",
            "- Installing BiocVersion ...                    OK [built from source and cached in 1.6s]\n",
            "- Installing BiocGenerics ...                   OK [built from source and cached in 6.6s]\n",
            "- Installing crayon ...                         OK [copied from cache in 14s]\n",
            "- Installing S4Vectors ...                      OK [built from source and cached in 31s]\n",
            "- Installing IRanges ...                        OK [built from source and cached in 1.3m]\n",
            "- Installing zlibbioc ...                       OK [built from source and cached in 2.4s]\n",
            "- Installing XVector ...                        OK [built from source and cached in 39s]\n",
            "- Installing curl ...                           OK [copied from cache in 22s]\n",
            "- Installing jsonlite ...                       OK [copied from cache in 24s]\n",
            "- Installing mime ...                           OK [copied from cache in 14s]\n",
            "- Installing sys ...                            OK [copied from cache in 15s]\n",
            "- Installing askpass ...                        OK [copied from cache in 16s]\n",
            "- Installing openssl ...                        OK [copied from cache in 25s]\n",
            "- Installing R6 ...                             OK [copied from cache in 15s]\n",
            "- Installing httr ...                           OK [copied from cache in 31s]\n",
            "- Installing UCSC.utils ...                     OK [built from source and cached in 5.3s]\n",
            "- Installing GenomeInfoDbData ...               OK [built from source and cached in 1.5s]\n",
            "- Installing GenomeInfoDb ...                   OK [built from source and cached in 37s]\n",
            "- Installing Biostrings ...                     OK [built from source and cached in 1.0m]\n",
            "- Installing DBI ...                            OK [installed binary and cached in 0.75s]\n",
            "- Installing Formula ...                        OK [installed binary and cached in 0.46s]\n",
            "- Installing cli ...                            OK [installed binary and cached in 0.6s]\n",
            "- Installing glue ...                           OK [installed binary and cached in 0.35s]\n",
            "- Installing rlang ...                          OK [installed binary and cached in 0.52s]\n",
            "- Installing lifecycle ...                      OK [installed binary and cached in 0.46s]\n",
            "- Installing gtable ...                         OK [installed binary and cached in 0.56s]\n",
            "- Installing isoband ...                        OK [installed binary and cached in 0.39s]\n",
            "- Installing MASS ...                           OK [built from source and cached in 11s]\n",
            "- Installing lattice ...                        OK [built from source and cached in 15s]\n",
            "- Installing Matrix ...                         OK [built from source and cached in 4.3m]\n",
            "- Installing nlme ...                           OK [built from source and cached in 23s]\n",
            "- Installing mgcv ...                           OK [built from source and cached in 1.3m]\n",
            "- Installing farver ...                         OK [installed binary and cached in 0.36s]\n",
            "- Installing labeling ...                       OK [installed binary and cached in 0.3s]\n",
            "- Installing RColorBrewer ...                   OK [installed binary and cached in 0.33s]\n",
            "- Installing viridisLite ...                    OK [installed binary and cached in 0.33s]\n",
            "- Installing scales ...                         OK [installed binary and cached in 0.86s]\n",
            "- Installing fansi ...                          OK [installed binary and cached in 0.46s]\n",
            "- Installing magrittr ...                       OK [installed binary and cached in 0.43s]\n",
            "- Installing utf8 ...                           OK [installed binary and cached in 0.48s]\n",
            "- Installing vctrs ...                          OK [installed binary and cached in 0.8s]\n",
            "- Installing pillar ...                         OK [installed binary and cached in 0.65s]\n",
            "- Installing pkgconfig ...                      OK [installed binary and cached in 0.3s]\n",
            "- Installing tibble ...                         OK [installed binary and cached in 0.75s]\n",
            "- Installing withr ...                          OK [installed binary and cached in 0.33s]\n",
            "- Installing ggplot2 ...                        OK [installed binary and cached in 1.1s]\n",
            "- Installing cluster ...                        OK [built from source and cached in 9.5s]\n",
            "- Installing rpart ...                          OK [built from source and cached in 6.9s]\n",
            "- Installing nnet ...                           OK [built from source and cached in 4.8s]\n",
            "- Installing foreign ...                        OK [built from source and cached in 7.0s]\n",
            "- Installing gridExtra ...                      OK [installed binary and cached in 0.57s]\n",
            "- Installing data.table ...                     OK [installed binary and cached in 0.49s]\n",
            "- Installing stringi ...                        OK [installed binary and cached in 0.76s]\n",
            "- Installing stringr ...                        OK [installed binary and cached in 0.9s]\n",
            "- Installing evaluate ...                       OK [installed binary and cached in 0.45s]\n",
            "- Installing xfun ...                           OK [installed binary and cached in 0.32s]\n",
            "- Installing highr ...                          OK [installed binary and cached in 0.31s]\n",
            "- Installing knitr ...                          OK [installed binary and cached in 0.42s]\n",
            "- Installing backports ...                      OK [installed binary and cached in 0.32s]\n",
            "- Installing checkmate ...                      OK [installed binary and cached in 0.42s]\n",
            "- Installing base64enc ...                      OK [installed binary and cached in 0.29s]\n",
            "- Installing digest ...                         OK [installed binary and cached in 0.34s]\n",
            "- Installing fastmap ...                        OK [installed binary and cached in 0.29s]\n",
            "- Installing htmltools ...                      OK [installed binary and cached in 0.5s]\n",
            "- Installing cachem ...                         OK [installed binary and cached in 0.5s]\n",
            "- Installing jquerylib ...                      OK [installed binary and cached in 0.54s]\n",
            "- Installing memoise ...                        OK [installed binary and cached in 0.7s]\n",
            "- Installing fs ...                             OK [installed binary and cached in 0.5s]\n",
            "- Installing rappdirs ...                       OK [installed binary and cached in 0.44s]\n",
            "- Installing sass ...                           OK [installed binary and cached in 0.87s]\n",
            "- Installing bslib ...                          OK [installed binary and cached in 0.94s]\n",
            "- Installing fontawesome ...                    OK [installed binary and cached in 0.53s]\n",
            "- Installing tinytex ...                        OK [installed binary and cached in 0.36s]\n",
            "- Installing rmarkdown ...                      OK [installed binary and cached in 1.0s]\n",
            "- Installing htmlwidgets ...                    OK [installed binary and cached in 0.54s]\n",
            "- Installing rstudioapi ...                     OK [installed binary and cached in 0.32s]\n",
            "- Installing htmlTable ...                      OK [installed binary and cached in 0.74s]\n",
            "- Installing viridis ...                        OK [installed binary and cached in 1.1s]\n",
            "- Installing colorspace ...                     OK [installed binary and cached in 0.48s]\n",
            "- Installing Hmisc ...                          OK [installed binary and cached in 1.7s]\n",
            "- Installing Rcpp ...                           OK [installed binary and cached in 1.1s]\n",
            "- Installing generics ...                       OK [installed binary and cached in 0.3s]\n",
            "- Installing tidyselect ...                     OK [installed binary and cached in 0.56s]\n",
            "- Installing dplyr ...                          OK [installed binary and cached in 0.97s]\n",
            "- Installing WikipediR ...                      OK [installed binary and cached in 0.39s]\n",
            "- Installing assertthat ...                     OK [installed binary and cached in 0.31s]\n",
            "- Installing ratelimitr ...                     OK [installed binary and cached in 0.31s]\n",
            "- Installing purrr ...                          OK [installed binary and cached in 0.59s]\n",
            "- Installing clipr ...                          OK [installed binary and cached in 0.46s]\n",
            "- Installing hms ...                            OK [installed binary and cached in 0.86s]\n",
            "- Installing bit ...                            OK [installed binary and cached in 0.51s]\n",
            "- Installing bit64 ...                          OK [installed binary and cached in 0.66s]\n",
            "- Installing cpp11 ...                          OK [installed binary and cached in 0.31s]\n",
            "- Installing tzdb ...                           OK [installed binary and cached in 0.37s]\n",
            "- Installing prettyunits ...                    OK [installed binary and cached in 0.29s]\n",
            "- Installing progress ...                       OK [installed binary and cached in 0.61s]\n",
            "- Installing vroom ...                          OK [installed binary and cached in 0.88s]\n",
            "- Installing readr ...                          OK [installed binary and cached in 0.82s]\n",
            "- Installing lazyeval ...                       OK [installed binary and cached in 0.3s]\n",
            "- Installing rex ...                            OK [installed binary and cached in 0.32s]\n",
            "- Installing WikidataQueryServiceR ...          OK [installed binary and cached in 0.32s]\n",
            "- Installing pbapply ...                        OK [installed binary and cached in 0.36s]\n",
            "- Installing WikidataR ...                      OK [installed binary and cached in 1.9s]\n",
            "- Installing XML ...                            OK [installed binary and cached in 0.89s]\n",
            "- Installing ape ...                            OK [installed binary and cached in 0.64s]\n",
            "- Installing blob ...                           OK [installed binary and cached in 0.52s]\n",
            "- Installing xml2 ...                           OK [installed binary and cached in 0.49s]\n",
            "- Installing triebeard ...                      OK [installed binary and cached in 0.54s]\n",
            "- Installing urltools ...                       OK [installed binary and cached in 0.61s]\n",
            "- Installing httpcode ...                       OK [installed binary and cached in 0.29s]\n",
            "- Installing crul ...                           OK [installed binary and cached in 0.37s]\n",
            "- Installing bold ...                           OK [built from source and cached in 5.5s]\n",
            "- Installing tidyr ...                          OK [installed binary and cached in 0.95s]\n",
            "- Installing broom ...                          OK [installed binary and cached in 0.99s]\n",
            "- Installing ps ...                             OK [installed binary and cached in 0.31s]\n",
            "- Installing processx ...                       OK [installed binary and cached in 0.35s]\n",
            "- Installing callr ...                          OK [installed binary and cached in 0.35s]\n",
            "- Installing rematch ...                        OK [installed binary and cached in 0.3s]\n",
            "- Installing cellranger ...                     OK [installed binary and cached in 0.35s]\n",
            "- Installing commonmark ...                     OK [installed binary and cached in 0.33s]\n",
            "- Installing conflicted ...                     OK [installed binary and cached in 0.48s]\n",
            "- Installing dbplyr ...                         OK [installed binary and cached in 1.2s]\n",
            "- Installing dtplyr ...                         OK [installed binary and cached in 1.5s]\n",
            "- Installing duckdb ...                         OK [installed binary and cached in 2.1s]\n",
            "- Installing duckdbfs ...                       OK [installed binary and cached in 0.37s]\n",
            "- Installing fastmatch ...                      OK [installed binary and cached in 0.41s]\n",
            "- Installing forcats ...                        OK [installed binary and cached in 0.59s]\n",
            "- Installing gargle ...                         OK [installed binary and cached in 0.62s]\n",
            "- Installing ggrepel ...                        OK [installed binary and cached in 1.5s]\n",
            "- Installing litedown ...                       OK [installed binary and cached in 0.35s]\n",
            "- Installing markdown ...                       OK [installed binary and cached in 0.29s]\n",
            "- Installing png ...                            OK [installed binary and cached in 0.32s]\n",
            "- Installing jpeg ...                           OK [installed binary and cached in 0.33s]\n",
            "- Installing gridtext ...                       OK [installed binary and cached in 0.88s]\n",
            "- Installing ggtext ...                         OK [installed binary and cached in 2.3s]\n",
            "- Installing uuid ...                           OK [installed binary and cached in 0.45s]\n",
            "- Installing googledrive ...                    OK [installed binary and cached in 0.94s]\n",
            "- Installing ids ...                            OK [installed binary and cached in 0.3s]\n",
            "- Installing rematch2 ...                       OK [installed binary and cached in 0.68s]\n",
            "- Installing googlesheets4 ...                  OK [installed binary and cached in 0.97s]\n",
            "- Installing haven ...                          OK [installed binary and cached in 0.82s]\n",
            "- Installing rprojroot ...                      OK [installed binary and cached in 0.31s]\n",
            "- Installing here ...                           OK [installed binary and cached in 0.3s]\n",
            "- Installing igraph ...                         OK [installed binary and cached in 0.71s]\n",
            "- Installing timechange ...                     OK [installed binary and cached in 0.31s]\n",
            "- Installing lubridate ...                      OK [installed binary and cached in 0.79s]\n",
            "- Installing snakecase ...                      OK [installed binary and cached in 0.43s]\n",
            "- Installing janitor ...                        OK [installed binary and cached in 1.4s]\n",
            "- Installing later ...                          OK [installed binary and cached in 0.45s]\n",
            "- Installing modelr ...                         OK [installed binary and cached in 0.9s]\n",
            "- Installing natserv ...                        OK [installed binary and cached in 0.81s]\n",
            "- Installing patchwork ...                      OK [installed binary and cached in 1.3s]\n",
            "- Installing quadprog ...                       OK [installed binary and cached in 0.29s]\n",
            "- Installing phangorn ...                       OK [installed binary and cached in 3.1s]\n",
            "- Installing plyr ...                           OK [installed binary and cached in 0.67s]\n",
            "- Installing systemfonts ...                    OK [installed binary and cached in 0.77s]\n",
            "- Installing textshaping ...                    OK [installed binary and cached in 0.75s]\n",
            "- Installing ragg ...                           OK [installed binary and cached in 0.79s]\n",
            "- Installing readxl ...                         OK [installed binary and cached in 0.36s]\n",
            "- Installing rentrez ...                        OK [installed binary and cached in 0.5s]\n",
            "- Installing reprex ...                         OK [installed binary and cached in 0.56s]\n",
            "- Installing rfishbase ...                      OK [installed binary and cached in 0.92s]\n",
            "- Installing solrium ...                        OK [installed binary and cached in 1.1s]\n",
            "- Installing ritis ...                          OK [installed binary and cached in 1.2s]\n",
            "- Installing rncl ...                           OK [installed binary and cached in 0.83s]\n",
            "- Installing rotl ...                           OK [installed binary and cached in 1.6s]\n",
            "- Installing rredlist ...                       OK [installed binary and cached in 0.88s]\n",
            "- Installing selectr ...                        OK [installed binary and cached in 0.62s]\n",
            "- Installing rvest ...                          OK [installed binary and cached in 0.56s]\n",
            "- Installing zoo ...                            OK [installed binary and cached in 0.45s]\n",
            "- Installing worrms ...                         OK [installed binary and cached in 0.32s]\n",
            "- Installing wikitaxa ...                       OK [installed binary and cached in 0.31s]\n",
            "- Installing taxize ...                         OK [installed binary and cached in 0.95s]\n",
            "- Installing tidyverse ...                      OK [installed binary and cached in 0.31s]\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "In addition: There were 11 warnings (use warnings() to see them)\n"
            ]
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "---\n",
        "## Parameter Setting"
      ],
      "metadata": {
        "id": "w42DRnn4_Xm2"
      },
      "id": "w42DRnn4_Xm2"
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "min_read_count <- 1000 #Minimum number of reads in a sample to keep it"
      ],
      "metadata": {
        "id": "ggTnXKtZiQbs"
      },
      "id": "ggTnXKtZiQbs",
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "---\n",
        "## Read in Data"
      ],
      "metadata": {
        "id": "EuDsGt4t_jhC"
      },
      "id": "EuDsGt4t_jhC"
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "samples_with_barcodes <- read_tsv(str_c(here::here(),\n",
        "               'data/rb_input_barcode.txt',\n",
        "               sep = '/'),\n",
        "               show_col_types = FALSE) %>%\n",
        "  select(sample) %>%\n",
        "  separate_wider_delim(cols = sample,\n",
        "                       names = c(\"pool\", \"sample_id\"),\n",
        "                       delim = \"_\")\n",
        "\n",
        "sampling_metadata <- read_tsv(str_c(here::here(),\n",
        "                                    'data/Simons_MasterSheet_2017-11-10_2.tsv',\n",
        "                                   sep = '/'),\n",
        "                             show_col_types = FALSE, na = c('NA', '?', '')) %>%\n",
        "  clean_names() %>%\n",
        "  mutate(sample_id = str_replace_all(vial_label, '-', ''),\n",
        "         .keep = 'unused') %>%\n",
        "  inner_join(samples_with_barcodes,\n",
        "             by = 'sample_id')\n",
        "\n",
        "\n",
        "#there are two samples with sequencing (in abundance matrix/samples_with_barcodes) without any info in the metadata: 1949E & 1751A\n",
        "\n",
        "abundance_matrix <- read_csv(str_c(here::here(),\n",
        "                                   'output/rainbow_bridge/zotu_table.csv',\n",
        "                                   sep = '/'),\n",
        "                             show_col_types = FALSE)\n",
        "\n",
        "zotu_sequences <- Biostrings::readDNAStringSet(str_c(here::here(),\n",
        "                                   'output/rainbow_bridge/zotu_sequences.fasta',\n",
        "                                   sep = '/'))\n",
        "\n",
        "zotu_taxonomy <- abundance_matrix %>%\n",
        "  select(zotu, domain:species, taxid_rank) %>%\n",
        "  distinct()\n"
      ],
      "metadata": {
        "id": "MgBWJWXC_gAX"
      },
      "id": "MgBWJWXC_gAX",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "---\n",
        "## Data Summary"
      ],
      "metadata": {
        "id": "MRsV8AvPBi-Q"
      },
      "id": "MRsV8AvPBi-Q"
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "font_size <- 2\n",
        "plot_nzotu_nsamples <- joined_zotu_data %>%\n",
        "  summarise(n_samples = n_distinct(sample_id),\n",
        "            n_zotus = n_distinct(zotu),\n",
        "            n_reads = sum(n_reads),\n",
        "            .by = c(predator_species_name)) %>%\n",
        "  ggplot(aes(x = n_samples,\n",
        "             y= n_zotus)) +\n",
        "  geom_point() +\n",
        "  geom_text_repel(aes(label = predator_species_name),\n",
        "                  size = font_size,\n",
        "                  max.overlaps = 30,\n",
        "                  fontface=\"italic\") +\n",
        "  scale_x_continuous(transform = scales::log10_trans()) +\n",
        "  scale_y_continuous(transform = scales::log10_trans()) +\n",
        "  annotation_logticks() +\n",
        "  labs(x = \"No. of Samples (log<sub>10</sub>-scale)\",\n",
        "       y = \"No. of zOTUs (log<sub>10</sub>-scale)\",\n",
        "       subtitle = \"Number of zOTUs as a function of number of samples\") +\n",
        "  theme_classic(base_size = 10) +\n",
        "  theme(axis.title = element_markdown(),\n",
        "        panel.background = element_rect(colour = 'black'))\n",
        "\n",
        "plot_nreads_nsamples <- joined_zotu_data %>%\n",
        "  summarise(n_samples = n_distinct(sample_id),\n",
        "            n_zotus = n_distinct(zotu),\n",
        "            n_reads = sum(n_reads),\n",
        "            .by = c(predator_species_name)) %>%\n",
        "  ggplot(aes(x = n_samples,\n",
        "             y= n_reads)) +\n",
        "  geom_point() +\n",
        "  geom_text_repel(aes(label = predator_species_name),\n",
        "                  size = font_size,\n",
        "                  max.overlaps = 30,\n",
        "                  fontface=\"italic\") +\n",
        "  scale_x_continuous(transform = scales::log10_trans()) +\n",
        "  scale_y_continuous(transform = scales::log10_trans()) +\n",
        "  annotation_logticks() +\n",
        "  labs(x = \"No. of Samples (log<sub>10</sub>-scale)\",\n",
        "       y = \"No. of Reads (log<sub>10</sub>-scale)\",\n",
        "       subtitle = \"Number of reads as a function of number of samples\") +\n",
        "  theme_classic(base_size = 10) +\n",
        "  theme(axis.title = element_markdown(),\n",
        "        panel.background = element_rect(colour = 'black'))\n",
        "\n",
        "plot_nzotus_nreads <- joined_zotu_data %>%\n",
        "  summarise(n_samples = n_distinct(sample_id),\n",
        "            n_zotus = n_distinct(zotu),\n",
        "            n_reads = sum(n_reads),\n",
        "            .by = c(predator_species_name)) %>%\n",
        "  ggplot(aes(x = n_reads,\n",
        "             y= n_zotus)) +\n",
        "  geom_point() +\n",
        "  geom_text_repel(aes(label = predator_species_name),\n",
        "                  size = font_size,\n",
        "                  max.overlaps = 30,\n",
        "                  fontface=\"italic\") +\n",
        "  scale_x_continuous(transform = scales::log10_trans()) +\n",
        "  scale_y_continuous(transform = scales::log10_trans()) +\n",
        "  annotation_logticks() +\n",
        "  labs(x = \"No. of Reads (log<sub>10</sub>-scale)\",\n",
        "       y = \"No. of zOTUs (log<sub>10</sub>-scale)\",\n",
        "       subtitle = \"Number of zOTUs as a function of number of reads\") +\n",
        "  theme_classic(base_size = 10) +\n",
        "  theme(axis.title = element_markdown(),\n",
        "        panel.background = element_rect(colour = 'black'))\n",
        "\n",
        "plot_nzotu_nsamples + plot_nzotus_nreads + plot_nreads_nsamples + plot_spacer() +\n",
        "  plot_layout(nrow = 2, guides = 'collect', axes = 'collect')"
      ],
      "metadata": {
        "id": "mbkIgLo9Bn_x"
      },
      "id": "mbkIgLo9Bn_x",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "---\n",
        "## Data Cleaning\n",
        "Remove samples/zotus which don't meet prespecified criteria:\n",
        "\n",
        "\n",
        "1.   zOTU sequence must be 313 bp length\n",
        "2.   Remove zOTUs identified as the predator species\n",
        "3. Remove Predator species with less than 1,000 reads across all samples/zOTUs"
      ],
      "metadata": {
        "id": "gZy0xj0BBoU9"
      },
      "id": "gZy0xj0BBoU9"
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "msgs <- c()\n",
        "\n",
        "filter_predators <- function(data, min_reads){\n",
        "  out <- data %>%\n",
        "    filter(sum(n_reads) >= min_read_count,\n",
        "           .by = c(predator_species_name))\n",
        "\n",
        "  removed_preds <- anti_join(data, out,\n",
        "                             by = colnames(data)) %>%\n",
        "    pull(predator_species_name) %>% n_distinct()\n",
        "\n",
        "  msgs <<- c(msgs, paste0('\\n', 'Number of predator species with fewer than ',\n",
        "          scales::comma(min_reads), ' total reads: ', scales::comma(removed_preds)))\n",
        "  msgs <<- c(msgs, paste0('Predator species remaining: ', scales::comma(n_distinct(out$predator_species_name)), '\\n'))\n",
        "  out\n",
        "}\n",
        "\n",
        "filter_self <- function(data){\n",
        "  out <- data %>%\n",
        "    filter(species != predator_species_name)\n",
        "\n",
        "  removed_zotus <- anti_join(data, out,\n",
        "                             by = colnames(data)) %>%\n",
        "    pull(zotu) %>% n_distinct()\n",
        "\n",
        "  msgs <<- c(msgs, paste0('\\n', 'Number of zOTUs matching predator species: ', scales::comma(removed_zotus)))\n",
        "  msgs <<- c(msgs, paste0('zOTUs remaining: ', scales::comma(n_distinct(out$zotu)), '\\n'))\n",
        "\n",
        "  out\n",
        "}\n",
        "\n",
        "# Identify zotus with a non-standard length sequence\n",
        "zotus_to_keep <- names(zotu_sequences)[Biostrings::width(zotu_sequences) == 313] %>%\n",
        "  str_extract('Zotu[0-9]+')\n",
        "\n",
        "msgs <- c(msgs, paste0('\\n', 'Number of zOTUs not 313 bp long: ',\n",
        "                       scales::comma(length(zotu_sequences) - length(zotus_to_keep))))\n",
        "msgs <- c(msgs, paste0('zOTUs remaining: ', scales::comma(length(zotus_to_keep)), '\\n'))\n",
        "\n",
        "# Main processing\n",
        "predator_gut_contents <- joined_zotu_data %>%\n",
        "  filter(zotu %in% zotus_to_keep) %>%\n",
        "  filter_self() %>%\n",
        "  select(-species) %>%\n",
        "  filter_predators(min_read_count)\n",
        "\n",
        "# Finally, print all collected messages\n",
        "cat(paste(msgs, collapse = \"\\n\"))"
      ],
      "metadata": {
        "id": "L8BnKvU9BwG9"
      },
      "id": "L8BnKvU9BwG9",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "---\n",
        "## Post Cleaning Data Summary"
      ],
      "metadata": {
        "id": "9F6-pReOBwav"
      },
      "id": "9F6-pReOBwav"
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "font_size <- 2\n",
        "plot_nzotu_nsamples <- predator_gut_contents %>%\n",
        "  summarise(n_samples = n_distinct(sample_id),\n",
        "            n_zotus = n_distinct(zotu),\n",
        "            n_reads = sum(n_reads),\n",
        "            .by = c(predator_species_name)) %>%\n",
        "  ggplot(aes(x = n_samples,\n",
        "             y= n_zotus)) +\n",
        "  geom_point() +\n",
        "  geom_text_repel(aes(label = predator_species_name),\n",
        "                  size = font_size,\n",
        "                  max.overlaps = 30,\n",
        "                  fontface=\"italic\") +\n",
        "  scale_x_continuous(transform = scales::log10_trans()) +\n",
        "  scale_y_continuous(transform = scales::log10_trans()) +\n",
        "  annotation_logticks() +\n",
        "  labs(x = \"No. of Samples (log<sub>10</sub>-scale)\",\n",
        "       y = \"No. of zOTUs (log<sub>10</sub>-scale)\",\n",
        "       subtitle = \"Number of zOTUs as a function of number of samples\") +\n",
        "  theme_classic(base_size = 10) +\n",
        "  theme(axis.title = element_markdown(),\n",
        "        panel.background = element_rect(colour = 'black'))\n",
        "\n",
        "plot_nreads_nsamples <- predator_gut_contents %>%\n",
        "  summarise(n_samples = n_distinct(sample_id),\n",
        "            n_zotus = n_distinct(zotu),\n",
        "            n_reads = sum(n_reads),\n",
        "            .by = c(predator_species_name)) %>%\n",
        "  ggplot(aes(x = n_samples,\n",
        "             y= n_reads)) +\n",
        "  geom_point() +\n",
        "  geom_text_repel(aes(label = predator_species_name),\n",
        "                  size = font_size,\n",
        "                  max.overlaps = 30,\n",
        "                  fontface=\"italic\") +\n",
        "  scale_x_continuous(transform = scales::log10_trans()) +\n",
        "  scale_y_continuous(transform = scales::log10_trans()) +\n",
        "  annotation_logticks() +\n",
        "  labs(x = \"No. of Samples (log<sub>10</sub>-scale)\",\n",
        "       y = \"No. of Reads (log<sub>10</sub>-scale)\",\n",
        "       subtitle = \"Number of reads as a function of number of samples\") +\n",
        "  theme_classic(base_size = 10) +\n",
        "  theme(axis.title = element_markdown(),\n",
        "        panel.background = element_rect(colour = 'black'))\n",
        "\n",
        "plot_nzotus_nreads <- predator_gut_contents %>%\n",
        "  summarise(n_samples = n_distinct(sample_id),\n",
        "            n_zotus = n_distinct(zotu),\n",
        "            n_reads = sum(n_reads),\n",
        "            .by = c(predator_species_name)) %>%\n",
        "  ggplot(aes(x = n_reads,\n",
        "             y= n_zotus)) +\n",
        "  geom_point() +\n",
        "  geom_text_repel(aes(label = predator_species_name),\n",
        "                  size = font_size,\n",
        "                  max.overlaps = 30,\n",
        "                  fontface=\"italic\") +\n",
        "  scale_x_continuous(transform = scales::log10_trans()) +\n",
        "  scale_y_continuous(transform = scales::log10_trans()) +\n",
        "  annotation_logticks() +\n",
        "  labs(x = \"No. of Reads (log<sub>10</sub>-scale)\",\n",
        "       y = \"No. of zOTUs (log<sub>10</sub>-scale)\",\n",
        "       subtitle = \"Number of zOTUs as a function of number of reads\") +\n",
        "  theme_classic(base_size = 10) +\n",
        "  theme(axis.title = element_markdown(),\n",
        "        panel.background = element_rect(colour = 'black'))\n",
        "\n",
        "plot_nzotu_nsamples + plot_nzotus_nreads + plot_nreads_nsamples + plot_spacer() +\n",
        "  plot_layout(nrow = 2, guides = 'collect', axes = 'collect')"
      ],
      "metadata": {
        "id": "zNA5ixE6CD8r"
      },
      "id": "zNA5ixE6CD8r",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "---\n",
        "## Get Predator Information"
      ],
      "metadata": {
        "id": "34bLwx1sCEfC"
      },
      "id": "34bLwx1sCEfC"
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "get_fishbase_data <- function(species_df, species_col = 'species',\n",
        "                              datasets = c('species', 'ecology', 'estimate',\n",
        "                                           'morphology', 'reproduction', 'morphometrics')){\n",
        "  #Datasets: the rfishbase functions (quoted) to get information from\n",
        "\n",
        "  out <- species_df\n",
        "  for(func_name in datasets){\n",
        "\n",
        "    func <- get(func_name, envir = asNamespace(\"rfishbase\"))\n",
        "    search_result <- func(pull(species_df, species_col)) %>%\n",
        "      janitor::remove_empty(which = 'cols')\n",
        "\n",
        "    join_vec <- join_by(!!sym(species_col) == 'Species',\n",
        "                        !!!syms(intersect(colnames(out),\n",
        "                                          colnames(search_result))))\n",
        "\n",
        "    out <- out %>%\n",
        "      left_join(search_result,\n",
        "                by = join_vec) %>%\n",
        "      nest(!!sym(str_c('fishbase_search_', func_name)) := str_subset(colnames(search_result), '^Species$', negate = TRUE))\n",
        "  }\n",
        "\n",
        "  out\n",
        "}\n",
        "\n",
        "if(!file.exists(str_c(here::here(),\n",
        "                      'intermediate_files/predator_taxonomy.rds',\n",
        "                      sep = '/'))){\n",
        "  fishbase_db <- load_taxa() %>%\n",
        "    clean_names()\n",
        "\n",
        "  predator_taxonomy <- sampling_metadata %>%\n",
        "    mutate(predator_species_name = case_when(predator_species_name == \"Ophidion welshi\" ~ \"Ophidion josephi\",\n",
        "                                             TRUE ~ predator_species_name)) %>%\n",
        "    distinct(predator_species_name) %>%\n",
        "    pull(predator_species_name) %>%\n",
        "    tax_name(get = c(\"kingdom\", \"phylum\", \"class\", \"order\", \"family\", \"genus\", \"species\"),\n",
        "             db = \"ncbi\") %>%\n",
        "    as_tibble() %>%\n",
        "    left_join(fishbase_db %>% select(species, order_fishbase = order), by = \"species\") %>%\n",
        "    mutate(order = coalesce(order, order_fishbase),\n",
        "           .keep = 'unused') %>%\n",
        "    select(-db, -query) %>%\n",
        "    get_fishbase_data()\n",
        "}\n"
      ],
      "metadata": {
        "id": "LyKg_u7ECIgV"
      },
      "id": "LyKg_u7ECIgV",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "---\n",
        "## Write Cleaned files for future use\n",
        "TODO - CHANGE TO WRITE TO PERSISTENT STORAGE"
      ],
      "metadata": {
        "id": "wzahE76LCI46"
      },
      "id": "wzahE76LCI46"
    },
    {
      "cell_type": "code",
      "source": [
        "%%R\n",
        "dir.create(str_c(here::here(), 'intermediate_files', sep = '/'), showWarnings = FALSE)\n",
        "predator_gut_contents %>%\n",
        "  pivot_wider(names_from = 'zotu',\n",
        "              values_from = n_reads,\n",
        "              values_fill = 0L) %>%\n",
        "  write_csv(str_c(here::here(),\n",
        "                  'intermediate_files/filtered_zotu_counts.csv',\n",
        "                  sep = '/'))\n",
        "\n",
        "write_csv(zotu_taxonomy,\n",
        "          str_c(here::here(),\n",
        "                  'intermediate_files/zotu_taxonomy.csv',\n",
        "                  sep = '/'))\n",
        "\n",
        "write_rds(predator_taxonomy,\n",
        "          str_c(here::here(),\n",
        "                'intermediate_files/predator_taxonomy.rds',\n",
        "                sep = '/'))"
      ],
      "metadata": {
        "id": "m52-GmHBCOJA"
      },
      "id": "m52-GmHBCOJA",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#Cleanly unmount Google Drive\n",
        "drive.flush_and_unmount()"
      ],
      "metadata": {
        "id": "p2IQPEo_I_Pk"
      },
      "id": "p2IQPEo_I_Pk",
      "execution_count": null,
      "outputs": []
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}