---
title: "Alpha Diversity"
author: "TAMUCC - Genomics Core Laboratory"
format: html
editor: visual

execute: 
  echo: true
  warning: false
---

## Set-up Analysis Environment

```{r}
#| warning: false
#| code-summary: "Libraries"
#| code-fold: true

library(tidyverse)
library(ggnested)
library(ggtext)
library(patchwork)
```

## Read Filtered Data

```{r}
predator_gut_contents <- read_csv(str_c(here::here(), 
                                        'intermediate_files/filtered_zotu_counts.csv',
                                        sep = '/'),
                                  show_col_types = FALSE)

zotu_taxonomy <- read_csv(str_c(here::here(), 
                                'intermediate_files/zotu_taxonomy.csv',
                                sep = '/'),
                          show_col_types = FALSE)

predator_taxonomy <- read_rds(str_c(here::here(), 
                                    'intermediate_files/predator_taxonomy.rds',
                                    sep = '/'))
```

## Plot Predator Stomach Contents Composition

```{r}
sample_composition <- predator_gut_contents %>%
  pivot_longer(cols = starts_with('Zotu'),
               names_to = 'zotu',
               values_to = 'n_reads') %>%
  filter(n_reads > 0) %>%
  left_join(zotu_taxonomy,
            by = 'zotu') %>%
  mutate(across(where(is.character), 
                ~if_else(. == 'LCA_dropped', NA_character_, .))) %>%
  
  #Skip species for the sake of the legend
  mutate(lowest_level = case_when(#!is.na(species) ~ str_c('s_', species),
                                  !is.na(genus) ~ str_c('g_', genus, ' sp.'),
                                  !is.na(family) ~ str_c('f_', family),
                                  !is.na(order) ~ str_c('o_', order),
                                  !is.na(class) ~ str_c('c_', class),
                                  !is.na(phylum) ~ str_c('p_', phylum),
                                  !is.na(kingdom) ~ str_c('k_', kingdom),
                                  !is.na(domain) ~ str_c('d_', domain),
                                  TRUE ~ 'Unknown'),
         upper_level = case_when(phylum == 'Chordata' & !taxid_rank %in% c('domain', 'kingdom', 'phylum') ~ str_c('c_', class),
                                 TRUE ~ str_c('p_', phylum)),
         .after = species) %>%
  
  summarise(n_reads = sum(n_reads),
            zotu = unique(zotu) %>% str_c(collapse = '; '),
            .by = c(predator_species_name, upper_level, lowest_level))

sample_composition %>%
  left_join(predator_taxonomy,
            by = c('predator_species_name' = 'species')) %>%
  ggnested(aes(y = predator_species_name, x = n_reads, 
               main_group = upper_level, sub_group = lowest_level),
           legend_labeling = 'sub', legend_title = 'Lowest Taxonomic\nClassification',
           main_keys = TRUE, nested_aes = c("fill"), 
           gradient_type = 'both') +
  geom_col(position = 'fill') +
  scale_x_continuous(labels = scales::percent_format()) +
  guides(fill = guide_legend(ncol = 7)) + #ncol = 5
  facet_grid(order ~ ., scales = 'free_y', space = 'free_y', switch = 'y') +
  labs(y = NULL, 
       x = 'Relative Number of Reads (%)') +
  theme_classic(base_size = 12) +
  theme(axis.title.x = element_markdown(),
        axis.title.y = element_markdown(),
        panel.background = element_rect(colour = 'black'),
        legend.position = 'right',
        # legend.text = element_markdown(size = 2),
        # legend.key.size = unit(0.1, "line"),
        legend.key = element_blank(),
        strip.text.y = element_text(angle = 180), 
        strip.placement = "outside")

```
